File: 860.柠檬水找零.py
#
# @lc app=leetcode.cn id=860 lang=python3
#
# [860] 柠檬水找零
#

# @lc code=start
from typing import List
class Solution:
    def lemonadeChange(self, bills: List[int]) -> bool:
        counts = [0, 0]
        for i in bills:
            if i == 5:
                counts[0] += 1
            elif i == 10:
                if counts[0] < 1: return False
                counts[0] -= 1
                counts[1] += 1
            else:
                if counts[1] >= 1 and counts[0] >= 1:
                    counts[0] -= 1
                    counts[1] -= 1
                elif counts[0] >= 3:
                    counts[0] -= 3
                else: return False
        return True
# @lc code=end

File: 860.柠檬水找零.cpp
/*
 * @lc app=leetcode.cn id=860 lang=cpp
 *
 * [860] 柠檬水找零
 */

// @lc code=start
#include<vector>
using namespace std;
class Solution {
public:
    bool lemonadeChange(vector<int>& bills) {
        int c5 = 0, c10 = 0;
        for(auto i: bills){
            switch (i) {
                case 5:
                    ++c5;
                break;
                case 10:
                    if(c5){
                        --c5;
                        ++c10;
                    }
                    else return false;
                break;
                case 20:
                    if(c10 && c5){
                        --c5;
                        --c10;
                    }
                    else if(c5 >= 3) c5 -= 3;
                    else return false;
            }
        }
        return true;
    }
};
// @lc code=end

